src/System.Net.Security/src/System/Net/Security/SslState.cs

```diff
@@ -1346,23 +1358,30 @@ internal void FinishWrite()
 
             lock (this)
             {
-                HandleWriteCallback();
+                HandleQueuedCallback(ref _queuedWriteStateRequest);
             }
         }
 
-        private void HandleWriteCallback()
+        private void HandleQueuedCallback(ref object queuedStateRequest)
         {
-            object obj = _queuedWriteStateRequest;
-            _queuedWriteStateRequest = null;
+            object obj = queuedStateRequest;
+            if (obj == null)
+            {
+                return;
+            }
+            queuedStateRequest = null;
+                        
             switch (obj)
             {
-                case null:
-                    break;
                 case LazyAsyncResult lazy:
                     lazy.InvokeCallback();
                     break;
-                case TaskCompletionSource<int> tsc:
-                    tsc.SetResult(0);
+                case TaskCompletionSource<int> taskCompletionSource when taskCompletionSource.Task.AsyncState != null:
+                    Memory<byte> array = (Memory<byte>)taskCompletionSource.Task.AsyncState;
+                    taskCompletionSource.SetResult(CheckOldKeyDecryptedData(array));
+                    break;
+                case TaskCompletionSource<int> taskCompletionSource:
```

Comment:
How common is this?  Seems like this construction is going to result in multiple type checks, when we could instead do:
```C#
case TaskCompletionSource<int> taskCompletionSource:
    if (taskCompletionSource.Task.AsyncState != null)
    {
        ...
    }
    else
    {
        ...
    }
    break;
```